### ISSUE TITLE
[gpu] python 3.8.10 import and error

### ISSUE THREAD

**Issue by robrobSK** [microsoft/LightGBM#5261]
## Description

here the error when doing the test on python. able to complie lightgbm to run on windows 10(64bit) via cmake, mingw64 complier.

Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import lightgbm as lgb
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python\Python38_10-64\lib\site-packages\lightgbm\__init__.py", line 8, in <module>
    from .basic import Booster, Dataset, Sequence, register_logger
  File "C:\Python\Python38_10-64\lib\site-packages\lightgbm\basic.py", line 125, in <module>
    _LIB = _load_lib()
  File "C:\Python\Python38_10-64\lib\site-packages\lightgbm\basic.py", line 116, in _load_lib
    lib = ctypes.cdll.LoadLibrary(lib_path[0])
  File "C:\Python\Python38_10-64\lib\ctypes\__init__.py", line 451, in LoadLibrary
    return self._dlltype(name)
  File "C:\Python\Python38_10-64\lib\ctypes\__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
FileNotFoundError: Could not find module 'C:\Python\Python38_10-64\lib\site-packages\lightgbm\lib_lightgbm.dll' (or one of its dependencies). Try using the full path with constructor syntax.

## Reproducible example

C:\github_repos\LightGBM\python-package>setup.py install --mingw --gpu
running install
C:\Python\Python38_10-64\lib\site-packages\setuptools\command\install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
INFO:LightGBM:Starting to compile the library.
INFO:LightGBM:Starting to compile with CMake and MinGW.
running build
running build_py
running egg_info
writing lightgbm.egg-info\PKG-INFO
writing dependency_links to lightgbm.egg-info\dependency_links.txt
writing requirements to lightgbm.egg-info\requires.txt
writing top-level names to lightgbm.egg-info\top_level.txt
reading manifest file 'lightgbm.egg-info\SOURCES.txt'
reading manifest template 'MANIFEST.in'
no previously-included directories found matching 'build'
warning: no files found matching '*.txt'
warning: no files found matching '*.so' under directory 'lightgbm'
warning: no files found matching 'compile\cmake\IntegratedOpenCL.cmake'
warning: no files found matching '*.so' under directory 'compile'
warning: no files found matching '*.dll' under directory 'compile\Release'
warning: no files found matching '*.dll' under directory 'compile\windows\x64\DLL'
warning: no previously-included files matching '*.py[co]' found anywhere in distribution
warning: no previously-included files found matching 'compile\external_libs\compute\.git'
adding license file 'LICENSE'
writing manifest file 'lightgbm.egg-info\SOURCES.txt'
copying lightgbm\VERSION.txt -> build\lib\lightgbm
running install_lib
creating C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\basic.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\callback.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\compat.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\dask.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\engine.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\libpath.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\plotting.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\sklearn.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\VERSION.txt -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
copying build\lib\lightgbm\__init__.py -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
INFO:LightGBM:Installing lib_lightgbm from: ['C:\\github_repos\\LightGBM\\lib_lightgbm.dll', 'C:\\github_repos\\LightGBM\\python-package\\compile\\lib_lightgbm.dll']
copying C:\github_repos\LightGBM\lib_lightgbm.dll -> C:\Python\Python38_10-64\Lib\site-packages\lightgbm
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\basic.py to basic.cpython-38.pyc
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\callback.py to callback.cpython-38.pyc
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\compat.py to compat.cpython-38.pyc
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\dask.py to dask.cpython-38.pyc
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\engine.py to engine.cpython-38.pyc
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\libpath.py to libpath.cpython-38.pyc
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\plotting.py to plotting.cpython-38.pyc
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\sklearn.py to sklearn.cpython-38.pyc
byte-compiling C:\Python\Python38_10-64\Lib\site-packages\lightgbm\__init__.py to __init__.cpython-38.pyc
running install_egg_info
Copying lightgbm.egg-info to C:\Python\Python38_10-64\Lib\site-packages\lightgbm-3.3.2.99-py3.8.egg-info
running install_scripts

## Environment info

LightGBM version or commit hash:

Command(s) you used to install LightGBM
C:\github_repos\LightGBM\python-package>setup.py install --mingw --gpu

```shell

```

windows 10(64bit) Python 3.8.10 (64 bit)

**Comment by CWJ666** [comment]
In jupyter notebook, I also encountered this situation:

i found that when  "Anaconda3/lib/site-packages/lightgbm/lib_lightgbm.dll" was replaced by "Lightgbm/Release/lib_lightgbm.dll", "import lightgbm" has no error, but when i train a model, it raise  one error

then when  "Anaconda3/lib/site-packages/lightgbm/lib_lightgbm.dll" was replaced by "Lightgbm/python-package/compile/release/lib_lightgbm.dll", the error  disappeared!

**Comment by CWJ666** [comment]
I also encountered this situation：
>Python 3.8.13 (default, Mar 28 2022, 06:59:08) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>> import lightgbm
['D:\\ProgramData\\Anaconda3\\envs\\test\\lib\\site-packages\\lightgbm\\lib_lightgbm.dll']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\ProgramData\Anaconda3\envs\test\lib\site-packages\lightgbm\__init__.py", line 8, in <module>
    from .basic import Booster, Dataset, Sequence, register_logger
  File "D:\ProgramData\Anaconda3\envs\test\lib\site-packages\lightgbm\basic.py", line 126, in <module>
    _LIB = _load_lib()
  File "D:\ProgramData\Anaconda3\envs\test\lib\site-packages\lightgbm\basic.py", line 117, in _load_lib
    lib = ctypes.cdll.LoadLibrary(lib_path[0])
  File "D:\ProgramData\Anaconda3\envs\test\lib\ctypes\__init__.py", line 451, in LoadLibrary
    return self._dlltype(name)
  File "D:\ProgramData\Anaconda3\envs\test\lib\ctypes\__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
FileNotFoundError: Could not find module 'D:\ProgramData\Anaconda3\envs\test\lib\site-packages\lightgbm\lib_lightgbm.dll' (or one of its dependencies). Try using the full path with constructor syntax.

['D:\\ProgramData\\Anaconda3\\envs\\test\\lib\\site-packages\\lightgbm\\lib_lightgbm.dll'] is lib_path.

The compilation process is based on the [lightgbm gpu version for windows](https://lightgbm.readthedocs.io/en/latest/Installation-Guide.html#build-gpu-version) process:
>vs 2019, boost 1_71_0, cuda 11.2
>>python setup.py install --gpu



I have tried looking for a solution to the problem and I found this thing with dll dependency walker in lib_lightgbm.dll: my computer is missing many dll dependencies!

but when i `python setup.py install` and `import lightgbm`, the above problem disappeared!

this problem may be similar to #499. @guolinke please help.

**Comment by robrobSK** [comment]
i tried that before. not sure why cmake installation just cannot be registered.

C:\github_repos\LightGBM\python-package>python setup.py install --gpu
running install
C:\Python\Python38_10-64\lib\site-packages\setuptools\command\install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
INFO:LightGBM:Starting to compile the library.
INFO:LightGBM:Starting to compile with Visual Studio 17 2022 (x64).
INFO:LightGBM:Starting to compile with Visual Studio 16 2019 (x64).
Traceback (most recent call last):
  File "setup.py", line 96, in silent_call
    subprocess.check_call(cmd, stderr=log, stdout=log)
  File "C:\Python\Python38_10-64\lib\subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', 'C:\\github_repos\\LightGBM\\python-package\\build_cpp', '--target', '_lightgbm', '--config', 'Release']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 346, in <module>
    setup(name='lightgbm',
  File "C:\Python\Python38_10-64\lib\site-packages\setuptools\__init__.py", line 155, in setup
    return distutils.core.setup(**attrs)
  File "C:\Python\Python38_10-64\lib\site-packages\setuptools\_distutils\core.py", line 148, in setup
    return run_commands(dist)
  File "C:\Python\Python38_10-64\lib\site-packages\setuptools\_distutils\core.py", line 163, in run_commands
    dist.run_commands()
  File "C:\Python\Python38_10-64\lib\site-packages\setuptools\_distutils\dist.py", line 967, in run_commands
    self.run_command(cmd)
  File "C:\Python\Python38_10-64\lib\site-packages\setuptools\_distutils\dist.py", line 986, in run_command
    cmd_obj.run()
  File "setup.py", line 258, in run
    compile_cpp(use_mingw=self.mingw, use_gpu=self.gpu, use_cuda=self.cuda, use_cuda_exp=self.cuda_exp, use_mpi=self.mpi,
  File "setup.py", line 203, in compile_cpp
    silent_call(["cmake", "--build", str(build_dir), "--target", "_lightgbm", "--config", "Release"], raise_error=True,
  File "setup.py", line 100, in silent_call
    raise Exception("\n".join((error_msg, LOG_NOTICE)))
Exception: Please install CMake first
The full version of error log was saved into C:\Users\User\LightGBM_compilation.log

C:\github_repos\LightGBM\python-package>

**Comment by jameslamb** [comment]
> but when i `python setup.py install` and `import lightgbm`, the above problem disappeared!

When building the LightGBM Python package from source, it's important that you use

```shell
python setup.py install
```

and not

```shell
setup.py install
```

So that the first instance of executable `python` on `PATH` is found and used to build the package.

> not sure why cmake installation just cannot be registered

I don't know what you mean by "registered". This error means that you do not have `cmake` / `cmake.exe` available on `PATH`. Since it looks like you are using an Anaconda distribution of Python, try using `conda` to install `cmake`.

```shell
conda install -c conda-forge --yes cmake
```

**Comment by jameslamb** [comment]
I've closed #5335 as a duplicate of this issue. It contains some other details of a similar problem.
